counterparty-core/counterpartycore/lib/gas.py

```diff
@@ -0,0 +1,116 @@
+import math
+
+from counterpartycore.lib import config, database, ledger, util
+
+
+def initialise(db):
+    cursor = db.cursor()
+    # transaction_count
+    cursor.execute(
+        """CREATE TABLE IF NOT EXISTS transaction_count(
+            block_index INTEGER,
+            difficulty_period INTEGER,
+            transaction_id INTEGER,
+            count INTEGER)
+        """
+    )
+    database.create_indexes(
+        cursor,
+        "transaction_count",
+        [
+            ["block_index"],
+            ["difficulty_period", "transaction_id"],
+        ],
+    )
+
+
+def get_transaction_count_for_difficulty_period(db, transaction_id, difficulty_period):
+    cursor = db.cursor()
+    count = cursor.execute(
+        """
+        SELECT count FROM transaction_count
+        WHERE transaction_id = ? AND difficulty_period = ?
+        ORDER BY rowid DESC
+        LIMIT 1
+    """,
+        (transaction_id, difficulty_period),
+    ).fetchone()
+    if count is None:
+        return 0
+    return count["count"]
+
+
+def increment_counter(db, transaction_id, block_index):
+    difficulty_period = block_index // 2016
+    current_count = get_transaction_count_for_difficulty_period(
+        db, transaction_id, difficulty_period
+    )
+    new_count = current_count + 1
+
+    bindings = {
+        "block_index": block_index,
+        "difficulty_period": difficulty_period,
+        "transaction_id": transaction_id,
+        "count": new_count,
+    }
+    ledger.insert_record(db, "transaction_count", bindings, "INCREMENT_TRANSACTION_COUNT")
+
+
+def get_average_transactions(db, transaction_id, block_index):
+    previous_difficulty_period = (block_index // 2016) - 1
```

Comment:
so @warrenpuffett and I talked about this: it should be a rolling average over the last 2016 blocks, rather than aligned with the difficulty period

Reply:
make sense. done here https://github.com/CounterpartyXCP/counterparty-core/pull/2180/commits/01f3c321da474840dfa0057e5c6661d35f4a8852